framework/src/media/audio/audio_manager.c

```diff
@@ -146,44 +152,127 @@ static const struct audio_samprate_map_entry_s g_audio_samprate_entry[] = {
 static audio_manager_result_t get_supported_sample_rate(int fd, uint8_t *sample_type, audio_card_type_t type);
-static uint32_t get_closest_samprate(unsigned origin_samprate, audio_card_type_t type);
-static unsigned int resample_stream_in(audio_card_info_t *cur_card, void *data, unsigned int frames);
-static unsigned int resample_stream_out(audio_card_info_t *cur_card, void *data, unsigned int frames);
+static uint32_t get_closest_samprate(audio_handler_t *handler, unsigned origin_samprate);
+static unsigned int resample_stream_in(audio_handler_t *handler, void *data, unsigned int frames);
+static unsigned int resample_stream_out(audio_handler_t *handler, void *data, unsigned int frames);
 static audio_manager_result_t get_audio_volume(int fd, audio_config_t *config, audio_card_type_t card_type);
-static audio_manager_result_t set_audio_volume(audio_card_type_t type, uint8_t volume);
+static audio_manager_result_t set_audio_volume(audio_handler_t *handler, audio_card_type_t type, uint8_t volume);
 
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
-static audio_manager_result_t init_audio_stream(audio_card_type_t card_type)
+static audio_manager_result_t get_audio_card_in(audio_card_operation_t card_operation, audio_handler_t **handler)
 {
-	if (((card_type == INPUT) && (g_actual_audio_in_card_id >= 0)) || ((card_type == OUTPUT) && (g_actual_audio_out_card_id >= 0))) {
-		return AUDIO_MANAGER_SUCCESS;
-	}
+	audio_manager_result_t ret = AUDIO_MANAGER_CARD_NOT_READY;
+	static audio_card_info_t *in_audio_card_tail;
+	static audio_card_info_t *in_audio_card_head;
+	audio_card_info_t *curr;
+
+	switch (card_operation) {
+	case ENABLE:
+		*handler = (audio_card_info_t *)malloc(sizeof(audio_card_info_t));
+		medvdbg("In handler addr = %x\n", *handler);
+		if ((*handler) == NULL) {
+			meddbg("Malloc for an audio handler (in) is failed\n");
+			return AUDIO_MANAGER_FAIL;
+		}
 
-	/* check available cards and try to find all cards if there's no one */
-	if (find_audio_card(card_type) != AUDIO_MANAGER_SUCCESS) {
-		meddbg("Found no active %s audio card\n", (card_type == OUTPUT ? "output" : "input"));
-		return AUDIO_MANAGER_NO_AVAIL_CARD;
+		(*handler)->next = NULL;
+		if (g_num_found_in_cards == 0) {
+			in_audio_card_tail = *handler;
+			in_audio_card_head = in_audio_card_tail;
+		} else {
+			in_audio_card_tail->next = *handler;
+			in_audio_card_tail = in_audio_card_tail->next;
+		}
+
+		ret = AUDIO_MANAGER_SUCCESS;
+		break;
+
+	case FIND:
+		curr = in_audio_card_head;
+		while (curr != NULL) {
+			if ((curr->status == AUDIO_CARD_IDLE)
+				|| (curr->status == AUDIO_CARD_READY)
+				|| (curr->status == AUDIO_CARD_PAUSE)) {
+				*handler = curr;
+				ret = AUDIO_MANAGER_SUCCESS;
+				break;
+			}
+			curr = curr->next;
+		}
+		break;
+
+	case DESTROY:
```

Comment:
Do not need AUDIO_MANAGER_SUCCESS as return?

Reply:
Good point. Modified.